pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,406 @@
+	default:
+		return nil
+	}
+}
+
+func makeBaseDiff(old, new resource.PropertyValue) baseDiff {
+	oldPresent := isPresent(old)
+	newPresent := isPresent(new)
+	if !oldPresent {
+		if !newPresent {
+			return NoDiff
+		}
+
+		return Add
+	}
+	if !newPresent {
+		return Delete
+	}
+
+	if new.IsComputed() {
+		return Update
+	}
+
+	return Undecided
+}
+
+type (
+	detailedDiffKey string
+	propertyPath    resource.PropertyPath
+)
+
+func newPropertyPath(root resource.PropertyKey) propertyPath {
+	return propertyPath{string(root)}
+}
+
+func (k propertyPath) String() string {
+	return resource.PropertyPath(k).String()
+}
+
+func (k propertyPath) Key() detailedDiffKey {
+	return detailedDiffKey(k.String())
+}
+
+func (k propertyPath) append(subkey interface{}) propertyPath {
+	return append(k, subkey)
+}
+
+func (k propertyPath) Subpath(subkey string) propertyPath {
+	return k.append(subkey)
+}
+
+func (k propertyPath) Index(i int) propertyPath {
+	return k.append(i)
+}
+
+func (k propertyPath) IsReservedKey() bool {
+	leaf := k[len(k)-1]
+	return leaf == "__meta" || leaf == "__defaults"
+}
+
+func mapHasReplacements(m map[detailedDiffKey]*pulumirpc.PropertyDiff) bool {
+	for _, diff := range m {
+		if diff.GetKind() == pulumirpc.PropertyDiff_ADD_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_DELETE_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_UPDATE_REPLACE {
+			return true
+		}
+	}
+	return false
+}
+
+type detailedDiffer struct {
+	tfs shim.SchemaMap
+	ps  map[string]*SchemaInfo
+}
+
+func (differ detailedDiffer) propertyPathToSchemaPath(path propertyPath) walk.SchemaPath {
+	return PropertyPathToSchemaPath(resource.PropertyPath(path), differ.tfs, differ.ps)
+}
+
+func (differ detailedDiffer) getEffectiveType(path walk.SchemaPath) shim.ValueType {
+	tfs, ps, err := LookupSchemas(path, differ.tfs, differ.ps)
+
+	if err != nil || tfs == nil {
+		return shim.TypeInvalid
+	}
+
+	if IsMaxItemsOne(tfs, ps) {
+		return differ.getEffectiveType(path.Element())
+	}
+
+	return tfs.Type()
+}
+
+func (differ detailedDiffer) lookupSchemas(path propertyPath) (shim.Schema, *info.Schema, error) {
+	schemaPath := PropertyPathToSchemaPath(resource.PropertyPath(path), differ.tfs, differ.ps)
+	return LookupSchemas(schemaPath, differ.tfs, differ.ps)
+}
+
+func (differ detailedDiffer) isForceNew(pair propertyPath) bool {
+	// See pkg/cross-tests/diff_cross_test.go
+	// TestAttributeCollectionForceNew, TestBlockCollectionForceNew, TestBlockCollectionElementForceNew
+	// for a full case study of replacements in TF
+	tfs, ps, err := differ.lookupSchemas(pair)
+	if err != nil {
+		return false
+	}
+	if isForceNew(tfs, ps) {
+		return true
+	}
+
+	if len(pair) == 1 {
+		return false
+	}
+
+	parent := pair[:len(pair)-1]
+	tfs, ps, err = differ.lookupSchemas(parent)
+	if err != nil {
+		return false
+	}
+	if tfs.Type() != shim.TypeList && tfs.Type() != shim.TypeSet && tfs.Type() != shim.TypeMap {
```

Comment:
Does this need to worry about effective type?

Reply:
no, since this is mimicing what is happening in TF-land - the flattening should not affect the ForceNew-iness of a property as it is directly translated from TF.

Reply:
OK